grails-app/controllers/org/pih/warehouse/inventory/StockMovementController.groovy

```diff
@@ -283,6 +287,98 @@ class StockMovementController {
+        StockMovement stockMovement = stockMovementService.getStockMovement(params['stockMovement.id'])
+        if (!stockMovement) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'stockMovement.label', default: 'StockMovement'), params['stockMovement.id']])}"
+            redirect(action: "list")
+            return
+        }
+        Comment comment = Comment.get(params?.id)
+        if (!comment) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'comment.label', default: 'Comment'), params.id])}"
+            redirect(action: "show", id: stockMovement.id)
+            return
+        }
+
+        if (comment.sender.id != session.user.id) {
+            throw new UnsupportedOperationException("${g.message(code: 'errors.noPermissions.label')}")
+        }
+        requisitionService.deleteComment(comment, stockMovement.requisition)
+        flash.message = "${g.message(code: 'default.comments.deleted.label')}"
+        redirect(action: "show", id: stockMovement.id)
+    }
+
+    def saveComment = {
+        StockMovement stockMovement = stockMovementService.getStockMovement(params['stockMovement.id'])
+        if (!stockMovement) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'stockMovement.label', default: 'StockMovement'), params['stockMovement.id']])}"
+            redirect(action: "list")
+            return
+        }
+        Comment comment = new Comment(params)
+         if (comment.validate()) {
+             requisitionService.addCommentToRequisition(comment, stockMovement.requisition)
+             render(view: "addComment", model: [stockMovement: stockMovement, comment: comment])
+             return
+         }
+        flash.message = "${g.message(code: 'default.comments.created.label')}"
+        redirect(action: "show", id: stockMovement.id)
+    }
+
+    def updateComment = {
+        StockMovement stockMovement = stockMovementService.getStockMovement(params['stockMovement.id'])
+        if (!stockMovement) {
+            flash.message = "${g.message(code: 'default.not.found.message', args: [g.message(code: 'stockMovement.label', default: 'StockMovement'), params['stockMovement.id']])}"
+            redirect(action: "list")
+            return
+        }
+        Comment comment = Comment.get(params?.id)
+        if (comment.sender.id != session.user.id) {
+            throw new UnsupportedOperationException("${g.message(code: 'errors.noPermissions.label')}")
+        }
+        comment.properties = params
+        if (comment.validate()) {
+            requisitionService.saveComment(comment)
+            render(view: "addComment", model: [stockMovement: stockMovement, comment: comment])
+            return
+        }
+        flash.message = "${g.message(code: 'default.comments.updated.label')}"
+        redirect(action: "show", id: stockMovement.id)
```

Comment:
I don't understand this flow. If the comment is valid, save it. Cool, but there's no handling of if the comment is not saved properly. And then if the comment is not valid, we still say it has been updated?

Reply:
Also we need default namespace here unless we use it without the "comments" part.

```
flash.message = "${g.message(code: 'default.updated.message', args: [g.message(code: 'comment.label', default: 'Comment'), comment.id])}"
```